File: tier0/engine/companion_standins.py
"""THE COMPANION STAND-IN SEAM (QUARANTINED, `C.COMPANION_OVERHAUL`).

A STAND-IN IS NOT A POOL MEMBER. It is a whole Klee-only card, with its own
unique name, handed to Klee IN PLACE of one named Universal (Klee brief pick 6;
the approved Mondstadt workshop sec.1; R236 sec.3). Every rule the seam has
follows from that one sentence:

  * it never enters ANY pool on its own -- not the character-blind Universal
    pool (`tier05.rewards.companion_pool`), not the Personal reward share, not
    a shop slot, not the Featured Banner, not an event. The ids are absent from
    `C.MONDSTADT_OVERHAUL_POOL_IDS`, so `companion_roster_replacement` cannot
    return one and no surface can tier one;
  * it is reached at the HAND-OFF and nowhere else -- after a surface has
    rolled its rarity and picked its card, `hand_off` swaps the Universal for
    the stand-in if this character owns one. The candidate lists, the rng
    stream and the weights are untouched, so THE OFFER ODDS DO NOT MOVE: a
    stand-in is offered exactly as often as the Universal it replaces was;
  * every other character is handed the Universal and never sees the stand-in,
    because the swap is keyed on `(replaces, personal_pool)`.

THE MAP IS DERIVED FROM THE SHEET, never listed here. A row states which
Universal it replaces (`replaces:`) and who may be handed it (`personal_pool:`);
a second, hand-copied table would be a list somebody must remember to update
and would fail silently the day a row is renamed.

C# TWIN: `KleeMod.Powers.CompanionStandIns`, which holds the same pair table
(by TYPE, so the compiler owns the correspondence) and is called from the same
two mouths -- the reward slot and the shop channel.

FLAG OFF EVERY FUNCTION HERE IS A NO-OP, checked at the top of each rather than
assumed by its callers, which is what makes the byte-identity pin
(`tier0/tests/test_companion_standins.py`) a property of this module.

THE FIVE CARETAKERS' RULES also live here, for the reason the seam does: they
are the only rules that read a stand-in, and a new file is what keeps a
quarantined arm's whole behaviour greppable in one place.

  * Shaken, Not Purred (Diona) -- ONE-SHOT. "If a Bomb goes off this turn,
    gain 5 Block." A printed conditional with no ordering word is true whether
    the Bomb went off before the card or after it, so the card pays at once
    when one already has and otherwise arms a watcher for the rest of the turn.
  * I Got Your Back (Noelle) -- REPEATING. "Whenever a Mine goes off this
    turn, gain 4 Block." `Whenever` is forward-looking and it pays per Mine.
  * Cold-Blooded Strike (Kaeya) -- a MARKER. "This turn, Grounded counts
    nothing as having gone off." Grounded reads LAST turn's count at the start
    of the next one, so the marker survives to that roll and is spent there.
  * Lion's Fang, Fair Protector (Jean) -- a POWER, Grounded's shape with a
    card on it: at the start of your turn, if none of your Bombs went off last
    turn, gain its stacks in Block and draw one.
  * Front Row Seat (Barbara, R252) -- REPEATING, and every Bomb. "Whenever a
    Bomb goes off this turn, gain 3 Block": Noelle's card with the Mines-only
    clause taken off, so a Mine pays both.

"THIS TURN" IS THE ROUND, including the enemy's half, and that is not a
liberty: Klee's Mines go off when an ENEMY attacks, so a window that closed at
the end of the player's own turn would leave "whenever a Mine goes off this
turn" unable to fire at all. Every watcher is therefore cleared where the
explosion counters roll -- the start of the player's NEXT turn.
"""

from __future__ import annotations

from functools import lru_cache
from typing import TYPE_CHECKING

from tier0 import constants as C

if TYPE_CHECKING:                                   # pragma: no cover
    from tier0.engine.state import Card, CombatState

#: Diona's one-shot watcher. Stacks are the BLOCK it pays, so the row's own
#: printed number is on the row and the Prototype-stage upgrade rule moves it.
SHAKEN_NOT_PURRED = "mc_shaken_not_purred"
#: Noelle's repeating watcher. Stacks are the Block paid per Mine.
I_GOT_YOUR_BACK = "mc_i_got_your_back"
#: Kaeya's marker. Stacks are a FLAG -- the row says so and states its own
#: upgrade rather than letting the rule move a number nothing reads.
COLD_BLOODED = "mc_cold_blooded"
#: Jean's power. Stacks are the Block, Grounded's own grammar.
LIONS_FANG = "mc_lions_fang"
#: Barbara's repeating watcher (R252). Stacks are the Block paid per BOMB --
#: Noelle's grammar with her Mines-only clause taken off, which is the one line
#: that separates the two.
FRONT_ROW_SEAT = "mc_front_row_seat"

#: The three watchers, cleared together at the turn roll.
_WATCHERS = (SHAKEN_NOT_PURRED, I_GOT_YOUR_BACK, FRONT_ROW_SEAT)


# --- the sheet contract ------------------------------------------------------

def validate_row(card: "Card") -> None:
    """`replaces:` is prototype surface only, and it needs a `personal_pool:`.

    Run from `loader._validate_card_shape`, which every sheet row passes
    through -- shipped and staged alike -- so the two halves are checked in one
    place:

    IT IS PROTOTYPE SURFACE ONLY, the same rule `description:` and `plan:`
    carry and for the same reason. A stand-in is a quarantined arm's card; a
    shipped row claiming to replace another would be a second, permanent pool
    rule wearing a schema key, and R213 B's deletion rule could never reach it.

    IT CANNOT BE ANONYMOUS. `replaces:` without `personal_pool:` names no
    character to hand the card to, so it would be a row that replaces a
    Universal for everybody -- which is a pool REPLACEMENT, not a stand-in, and
    the arm already has one of those (`MONDSTADT_OVERHAUL_POOL_IDS`).
    """
    from tier0.content.loader import PROTOTYPE_ID_PREFIX

    if card.replaces is None:
        return
    if not card.id.startswith(PROTOTYPE_ID_PREFIX):
        raise ValueError(
            f"card {card.id!r}: `replaces:` is prototype surface only -- a "
            f"shipped row may not stand in for another card (ids on that "
            f"surface carry {PROTOTYPE_ID_PREFIX!r})")
    if not card.personal_pool:
        raise ValueError(
            f"card {card.id!r}: `replaces:` needs a `personal_pool:` -- a "
            "stand-in is handed to ONE character in place of a Universal, and "
            "a row that replaces a Universal for everybody is a pool "
            "replacement rather than a stand-in")


@lru_cache(maxsize=1)
def _replacements() -> dict[tuple[str, str], str]:
    """`(universal id, character id) -> stand-in id`, derived from the sheet.

    Memoized off `loader._prototype_index`, which is itself the memoized read
    of the surface; `loader.reset_caches` drops both together.
    """
    from tier0.content import loader

    return {(c.replaces, c.personal_pool): c.id
            for c in loader._prototype_index().values()
            if c.replaces is not None and c.personal_pool}


def hand_off(card_id: str, character_id: str | None) -> str:
    """THE HAND-OFF, and the whole seam in the sim.

    Every companion offer surface calls this on the id it is about to hand
    over, AFTER the pick: `tier05.rewards.roll_rewards` (the reward slot) and
    `tier05.shop.companion_offers` (both shop slots). Called on the picked card
    rather than on the candidate list on purpose -- the lists, the weights and
    the rng draws are then provably the Universal's own, so no offer's odds can
    move.

    Returns `card_id` unchanged for every character but the stand-in's, and for
    every build with the arm off.
    """
    if not C.COMPANION_OVERHAUL or character_id is None:
        return card_id
    return _replacements().get((card_id, character_id), card_id)


def standin_ids() -> tuple[str, ...]:
    """Every stand-in the surface declares, sorted. The test seam, and the
    derivation `C.COMPANION_STANDIN_IDS` is pinned against."""
    return tuple(sorted(_replacements().values()))


# --- the four caretakers' rules ---------------------------------------------

def roll_turn(state: "CombatState") -> None:
    """The turn boundary: spend Kaeya's marker, close both watchers.

    Called from `combat._player_turn` on the line under
    `klee_overhaul.roll_to`, which is where this arm's explosion counters roll,
    so "this turn" means the same span to the watchers and to the counters.

    THE MARKER IS SPENT HERE and not inside Grounded's own check, because
    Grounded may not be in the deck: a marker that only a Grounded in play
    could clear would sit on the player forever and blind a Grounded drafted
    three fights later.
    """
    if not C.COMPANION_OVERHAUL:
        return
    p = state.player
    state.mc_grounded_blind = bool(p.powers.pop(COLD_BLOODED, 0))
    for name in _WATCHERS:
        p.powers.pop(name, None)


def grounded_blind(state: "CombatState") -> bool:
    """Does Grounded see nothing this turn? Read once, by
    `klee_overhaul.turn_start_late`, and false on every flag-off tree."""
    return C.COMPANION_OVERHAUL and state.mc_grounded_blind


def note_explosion(state: "CombatState", is_mine: bool) -> None:
    """One explosion landed: pay whichever watcher is armed.

    Called from `klee_overhaul._explode`, beside `note_explosion`, rather than
    through `_notify_explosion`: that bus carries no Mine flag, and widening it
    for one card would put a stand-in's rule inside the Klee arm's own hook.

    BLOCK IS RAW, the argument `companion_overhaul_turn_start` makes for every
    power-sourced Block on this arm: neither block funnel may touch a gain the
    card that banked it is no longer the source of.
    """
    if not C.COMPANION_OVERHAUL:
        return
    p = state.player
    # Diona -- ONE-SHOT, popped as it pays.
    n = p.powers.pop(SHAKEN_NOT_PURRED, 0)
    if n:
        p.block += n
        state.emit("block", amount=n)
        state.emit("mc_shaken_not_purred", amount=n)
    # Noelle -- REPEATING, and MINES ONLY.
    if is_mine:
        n = p.powers.get(I_GOT_YOUR_BACK, 0)
        if n:
            p.block += n
            state.emit("block", amount=n)
            state.emit("mc_i_got_your_back", amount=n)
    # Barbara -- REPEATING, and EVERY Bomb (R252). Noelle's shape without her
    # Mines clause, which is the whole difference between the two cards: a
    # Mine is a Bomb, so Noelle's window is a subset of this one.
    n = p.powers.get(FRONT_ROW_SEAT, 0)
    if n:
        p.block += n
        state.emit("block", amount=n)
        state.emit("mc_front_row_seat", amount=n)


def on_played(state: "CombatState", card: "Card") -> None:
    """Diona's card, played on a turn a Bomb has ALREADY gone off, pays now.

    "If a Bomb goes off this turn, gain 5 Block" carries no ordering word, so
    the condition is about the TURN and not about what happens next; a watcher
    alone would print a card that reads true and does nothing. Called from
    `effects.resolve_card` after the body, so the watcher the body just applied
    is the stack this spends.
    """
    if not C.COMPANION_OVERHAUL:
        return
    if state.ko_set_off_this_turn <= 0:
        return
    p = state.player
    n = p.powers.pop(SHAKEN_NOT_PURRED, 0)
    if n:
        p.block += n
        state.emit("block", amount=n)
        state.emit("mc_shaken_not_purred", amount=n)


def turn_start(state: "CombatState") -> None:
    """Jean, Lion's Fang, Fair Protector -- Grounded's shape with a card on it.

    Called from the tail of `effects.companion_overhaul_turn_start`, which is
    where this arm's start-of-turn payouts live, and it is COMMUTATIVE with the
    three already there: it grants the player Block and a card, and none of
    them reads a value it writes.

    IT READS `ko_set_off_last_turn` DIRECTLY, so it agrees with Grounded by
    construction. It does NOT read Kaeya's blind: that card names Grounded, and
    a marker that quietly paid a second power would be a rule the player was
    never shown.
    """
    if not C.COMPANION_OVERHAUL:
        return
    p = state.player
    n = p.powers.get(LIONS_FANG, 0)
    if not n or state.ko_set_off_last_turn != 0:
        return
    p.block += n
    state.emit("block", amount=n)
    state.draw(C.MC_LIONS_FANG_DRAW)
    state.emit("extra_draw", amount=C.MC_LIONS_FANG_DRAW)
    state.emit("mc_lions_fang", amount=n)

File: klee-mod/KleeCode/Powers/Prototype/CompanionStandIns.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using BaseLib.Abstracts;
using KleeMod.Cards;
using KleeMod.Cards.Prototype.Generated;
using MegaCrit.Sts2.Core.Combat;
using MegaCrit.Sts2.Core.Commands;
using MegaCrit.Sts2.Core.Entities.Cards;
using MegaCrit.Sts2.Core.Entities.Creatures;
using MegaCrit.Sts2.Core.Entities.Players;
using MegaCrit.Sts2.Core.Entities.Powers;
using MegaCrit.Sts2.Core.GameActions.Multiplayer;
using MegaCrit.Sts2.Core.Models;
using MegaCrit.Sts2.Core.Models.Powers;
using MegaCrit.Sts2.Core.ValueProps;

namespace KleeMod.Powers;

/// <summary>
/// THE COMPANION STAND-IN SEAM (QUARANTINED, <c>COMPANION_OVERHAUL</c>).
///
/// A STAND-IN IS NOT A POOL MEMBER. It is a whole Klee-only card, with its own
/// unique name, handed to Klee IN PLACE of one named Universal (Klee brief pick
/// 6; the approved Mondstadt workshop sec.1; R236 sec.3). Everything below
/// follows from that one sentence:
///
///   * it never enters ANY pool on its own. The four types are absent from
///     <see cref="CompanionOverhaulRoster"/>, which is the ONE door
///     <see cref="CompanionPool.All"/> opens, so the reward slot, the shop and
///     the Featured Banner are all structurally unable to see one;
///   * it is reached at the HAND-OFF and nowhere else. Each offer surface calls
///     <see cref="HandOff"/> on the card it has already PICKED, so the
///     eligibility lists, the rarity roll and the weighted draw are the
///     Universal's own and THE OFFER ODDS DO NOT MOVE;
///   * every other character is handed the Universal, because the swap is
///     keyed on the stand-in's own <c>PersonalPool</c>.
///
/// SIM TWIN: <c>tier0.engine.companion_standins</c>, called from the same two
/// mouths (<c>tier05.rewards.roll_rewards</c> and
/// <c>tier05.shop.companion_offers</c>).
///
/// THE PAIR TABLE IS LISTED BY TYPE, the same asymmetry
/// <see cref="CompanionOverhaulRoster"/> argues for its own list: a deleted row
/// takes its class with it and this file stops building, where a table of id
/// strings would fail silently the day a row is renamed. The sim derives the
/// same map from the sheet's <c>replaces:</c> key, and
/// <c>tier0/tests/test_companion_standins.py</c> pins the two against each
/// other by id.
///
/// PUBLIC rather than internal, for the reason <c>ProtoBombPower.Charges</c>
/// gives: KleeTests is a separate assembly, the decision <see cref="HandOffTo"/>
/// takes is the whole of this seam, and the alternative was an
/// <c>InternalsVisibleTo</c> nothing else in this mod needs or an IL-shape
/// assertion standing in for the decision itself -- which is exactly the
/// substitution that let the defect below ship.
/// </summary>
public static class CompanionStandIns
{
    /// <summary>
    /// The pairs, Universal -> stand-in, in the sheet's own order. Klee's four
    /// caretakers (R236 sec.3): Diona, Noelle, Kaeya, Jean.
    ///
    /// CACHED LAZILY for <see cref="CompanionOverhaulRoster"/>'s reason, which
    /// is the EB-194 lesson: <c>ModelDb.Card&lt;T&gt;()</c> throws until the
    /// models are built, and a static constructor that throws poisons its type
    /// for the life of the process.
    /// </summary>
    private static IReadOnlyList<(CardModel Universal, CardModel StandIn)>? _pairs;

    private static IReadOnlyList<(CardModel, CardModel)> Pairs() =>
        _pairs ??= new (CardModel, CardModel)[]
        {
            (ModelDb.Card<ProtoMcDionaIcyPaws>(),
             ModelDb.Card<ProtoMcDionaShakenNotPurred>()),
            (ModelDb.Card<ProtoMcNoelleBreastplate>(),
             ModelDb.Card<ProtoMcNoelleIGotYourBack>()),
            (ModelDb.Card<ProtoMcKaeyaFrostgnaw>(),
             ModelDb.Card<ProtoMcKaeyaColdBloodedStrike>()),
            (ModelDb.Card<ProtoMcJeanDandelionBreeze>(),
             ModelDb.Card<ProtoMcJeanLionsFang>()),
        };

    /// <summary>Test seam: forget the cache. The mod never calls it.</summary>
    internal static void ResetAll() => _pairs = null;

    /// <summary>
    /// THE HAND-OFF, and the whole seam in this engine.
    ///
    /// Called on a card a surface has already picked, immediately before it is
    /// instantiated for the player: <see cref="CompanionSlot.Roll"/> (the
    /// fourth reward slot) and <c>MerchantCompanionSlots.AddSlot</c> (both shop
    /// slots). Returns <paramref name="picked"/> unchanged for every character
    /// but the stand-in's own, and with the arm off.
    ///
    /// THE BANNER IS NOT A MOUTH, deliberately. It decides WHICH five-stars are
    /// featured, and it decides that about Universals; a stand-in carries a
    /// <c>PersonalPool</c>, which the banner's roster excludes by the same rule
    /// that keeps Klee's Personals off it.
    /// </summary>
    internal static CardModel HandOff(CardModel picked, Player player)
    {
        if (!CompanionOverhaul.Enabled) return picked;
        return HandOffTo(picked, CompanionPool.CharacterId(player), Pairs());
    }

    /// <summary>
    /// THE DECISION, with the pair table handed in instead of resolved, and
    /// the only reason it is a second method is that the PINS could not reach
    /// the first: <see cref="Pairs"/> goes through <c>ModelDb</c> and
    /// <see cref="HandOff"/> takes a <c>Player</c>, both outside the headless
    /// boundary (KleeTests/README), so the seam's own rule had no C# pin at all
    /// -- which is how it shipped broken. `CompanionStandInHandOffTests` calls
    /// THIS with two cards it constructed itself.
    ///
    /// THE COMPARISON IS THE WHOLE RULE, and it is a STRING one: the stand-in's
    /// <c>PersonalPool</c> must BE the character id
    /// <see cref="CompanionPool.CharacterId"/> returns. It was not. The codegen
    /// emitted a Python list repr (<c>"['klee']"</c>) for a row that spells
    /// `personal_pool:` as a one-member list, so this loop matched the pair,
    /// failed the second test and handed Klee the Universal at both mouths --
    /// silently, and in the engine the player plays, while the sim swapped
    /// correctly the whole time (`tools/gen_klee_cards.personal_pool_id` is the
    /// fix and `tier0.engine.state.Card.from_dict` the twin it was missing).
    /// </summary>
    public static CardModel HandOffTo(
        CardModel picked, string? characterId,
        IReadOnlyList<(CardModel Universal, CardModel StandIn)> pairs)
    {
        if (!CompanionOverhaul.Enabled) return picked;
        if (characterId == null) return picked;
        foreach (var (universal, standIn) in pairs)
        {
            if (!ReferenceEquals(universal, picked)) continue;
            if ((standIn as ICompanionCard)?.PersonalPool != characterId)
            {
                continue;
            }
            return standIn;
        }
        return picked;
    }

    // ---- the four caretakers' rules -------------------------------------

    /// <summary>
    /// One explosion landed: pay whichever this-turn watcher is armed.
    /// <c>companion_standins.note_explosion</c>'s twin, called from
    /// <c>ProtoBombPower.Explode</c> beside <c>NoteExplosion</c>.
    ///
    /// NOT ON THE EXPLOSION BUS (<c>IProtoExplosionListener</c>), and that is
    /// the one design note here: the bus carries no Mine flag, and Noelle's
    /// card is about Mines. Widening the Klee arm's own interface for one
    /// companion stand-in would put this arm's rule inside that one.
    /// </summary>
    internal static async Task OnExplosion(
        PlayerChoiceContext choiceContext, Creature applier, bool isMine)
    {
        if (!CompanionOverhaul.Enabled) return;
        foreach (var power in applier.Powers.ToList())
        {
            switch (power)
            {
                case ShakenNotPurredPower shaken:
                    await shaken.Pay();
                    break;
                case IGotYourBackPower back when isMine:
                    await back.Pay();
                    break;
            }
        }
    }

    /// <summary>
    /// Does Grounded see nothing this turn? Kaeya's Cold-Blooded Strike, read
    /// by <c>GroundedPower</c> and by nothing else -- the card names Grounded,
    /// so the blind is a READ here rather than a write to the explosion
    /// counter, which Jean's stand-in also reads.
    /// </summary>
    internal static bool GroundedBlind(Creature owner) =>
        CompanionOverhaul.Enabled && StandInLedger.For(owner).GroundedBlind;
}

/// <summary>
/// THE STAND-INS' TURN BOUNDARY, and it holds exactly one fact: whether
/// Grounded is blind this turn.
///
/// A SECOND LEDGER BESIDE <see cref="KleeOverhaulLedger"/>, not a field on it,
/// because the two arms ship independently: the Klee overhaul's counters are
/// rule 7's and exist with the companion arm off, and a companion stand-in's
/// marker must not be a field the Klee arm carries in a build that cannot
/// reach one. Same shape, same argument, same self-correcting round stamp --
/// see that class's header for why the roll is on READ rather than on a hook.
///
/// THE ROLL IS THE SPEND. Kaeya's card is played during round N and Grounded
/// asks at the start of round N+1, so the marker has to survive the enemy's
/// half; the roll reads it once at the boundary and caches the answer for the
/// round, which is what makes the read order-independent among the powers that
/// fire on one broadcast.
/// </summary>
internal sealed class StandInLedger
{
    private static object? _combat;
    private static readonly Dictionary<Creature, StandInLedger> _byOwner = new();

    private int _round = -1;

    /// <summary>Grounded's blind, for the round this ledger is rolled to.</summary>
    public bool GroundedBlind { get; private set; }

    public static StandInLedger For(Creature owner)
    {
        var combat = (object?)owner.CombatState;
        if (!ReferenceEquals(_combat, combat))
        {
            _combat = combat;
            _byOwner.Clear();
        }
        if (!_byOwner.TryGetValue(owner, out var ledger))
        {
            ledger = new StandInLedger();
            _byOwner[owner] = ledger;
        }
        ledger.RollTo(owner.CombatState?.RoundNumber ?? 0, owner);
        return ledger;
    }

    /// <summary>Test seam: forget everything. The mod never calls it.</summary>
    public static void ResetAll()
    {
        _combat = null;
        _byOwner.Clear();
    }

    /// <summary>Roll to <paramref name="round"/>. Public to the pins so a turn
    /// boundary can be exercised without a combat.</summary>
    public void RollTo(int round, Creature owner)
    {
        if (round == _round) return;
        // The marker is READ off the creature here rather than pushed by the
        // card, so the answer does not depend on an apply-time hook firing --
        // and a jump of more than one round means the owner had no turn in
        // between, which is honestly not blind.
        GroundedBlind = round == _round + 1
                        && owner.Powers.Any(p => p is ColdBloodedPower);
        _round = round;
    }
}

/// <summary>
/// Diona, Shaken, Not Purred: "Gain 6 Block. Apply Cryo twice. If a Bomb goes
/// off this turn, gain 5 Block."
///
/// ONE-SHOT, and the stack is the BLOCK it pays -- the row's own printed
/// number, so the Prototype-stage upgrade rule moves it like any other.
///
/// A CONDITIONAL WITH NO ORDERING WORD IS TRUE BOTH WAYS. "If a Bomb goes off
/// this turn" says nothing about before or after, so the generated card pays at
/// once when one already has (<c>KleeOverhaulLedger.SetOffThisTurn</c>, read in
/// <see cref="AfterCardPlayed"/>) and otherwise waits for the first explosion of
/// the turn. Either way the promise is kept once.
/// </summary>
public sealed class ShakenNotPurredPower : PowerModel, ILocalizationProvider
{
    public List<(string, string)>? Localization => new()
    {
        ("title", "Shaken, Not Purred"),
        ("description",
            "The next time one of your [gold]Bombs[/gold] goes off this turn, "
          + "gain [blue]{Amount}[/blue] [gold]Block[/gold]."),
    };

    public override PowerType Type => PowerType.Buff;

    public override PowerStackType StackType => PowerStackType.Counter;

    /// <summary>Pay and go. Called by <see cref="CompanionStandIns.OnExplosion"/>
    /// on any explosion, Mine or not -- a Mine is a Bomb.</summary>
    internal async Task Pay()
    {
        if (Owner == null || Amount <= 0) return;
        // NC-11: power-sourced BLOCK stays raw; only power-sourced DAMAGE runs
        // the pipeline (NC-1).
        await CreatureCmd.GainBlock(Owner, Amount, ValueProp.Unpowered, null,
                                    fast: true);
        await PowerCmd.Remove(this);
    }

    /// <summary>The card was played on a turn a Bomb had ALREADY gone off: the
    /// condition is about the turn, so it is already true.</summary>
    public override async Task AfterCardPlayed(
        PlayerChoiceContext choiceContext, CardPlay cardPlay)
    {
        if (Owner == null) return;
        if (KleeOverhaulLedger.For(Owner).SetOffThisTurn <= 0) return;
        await Pay();
    }

    /// <summary>"This turn" ends where the arm's counters roll -- the start of
    /// the owner's next turn, which leaves the enemy's half inside the window
    /// on purpose (a Mine goes off when an ENEMY attacks).</summary>
    public override async Task AfterPlayerTurnStart(
        PlayerChoiceContext choiceContext, Player player)
    {
        if (Owner == null || player.Creature != Owner) return;
        await PowerCmd.Remove(this);
    }
}

/// <summary>
/// Noelle, I Got Your Back: "Gain 6 Block. Whenever a Mine goes off this turn,
/// gain 4 Block."
///
/// REPEATING, and MINES ONLY. "Whenever" is forward-looking and pays per Mine,
/// which is the one place this card differs from Diona's above it.
/// </summary>
public sealed class IGotYourBackPower : PowerModel, ILocalizationProvider
{
    public List<(string, string)>? Localization => new()
    {
        ("title", "I Got Your Back"),
        ("description",
            "Whenever one of your [gold]Mines[/gold] goes off this turn, gain "
          + "[blue]{Amount}[/blue] [gold]Block[/gold]."),
    };

    public override PowerType Type => PowerType.Buff;

    public override PowerStackType StackType => PowerStackType.Counter;

    internal async Task Pay()
    {
        if (Owner == null || Amount <= 0) return;
        await CreatureCmd.GainBlock(Owner, Amount, ValueProp.Unpowered, null,
                                    fast: true);
    }

    public override async Task AfterPlayerTurnStart(
        PlayerChoiceContext choiceContext, Player player)
    {
        if (Owner == null || player.Creature != Owner) return;
        await PowerCmd.Remove(this);
    }
}

/// <summary>
/// Kaeya, Cold-Blooded Strike: "Deal 8 damage. Apply Cryo. This turn, Grounded
/// counts nothing as having gone off."
///
/// A MARKER, and its stack is a flag rather than a number -- which is why the
/// sheet row states its own upgrade instead of letting the Prototype-stage
/// rule move an amount nothing reads.
///
/// IT ROLLS THE LEDGER BEFORE IT LEAVES, and that is the whole of why the read
/// is race-free. Both this power and <c>GroundedPower</c> fire on
/// <c>AfterPlayerTurnStart</c>, whose listener order is not guaranteed; asking
/// <see cref="CompanionStandIns.GroundedBlind"/> first forces
/// <see cref="StandInLedger"/> to roll and CACHE the answer for the round while
/// this power is still on the creature, so Grounded reads the same true
/// whichever of the two ran first.
/// </summary>
public sealed class ColdBloodedPower : PowerModel, ILocalizationProvider
{
    public List<(string, string)>? Localization => new()
    {
        ("title", "Cold-Blooded"),
        ("description", "This turn, Grounded counts nothing as having gone off."),
    };

    public override PowerType Type => PowerType.Buff;

    public override PowerStackType StackType => PowerStackType.Counter;

    public override async Task AfterPlayerTurnStart(
        PlayerChoiceContext choiceContext, Player player)
    {
        if (Owner == null || player.Creature != Owner) return;
        CompanionStandIns.GroundedBlind(Owner);      // roll, and cache, first
        await PowerCmd.Remove(this);
    }
}

/// <summary>
/// Jean, Lion's Fang, Fair Protector: "At the start of your turn, if none of
/// your Bombs went off last turn, gain 8 Block and draw 1 card."
///
/// GROUNDED'S SHAPE WITH A CARD ON IT, and it reads the ledger the same way for
/// the same reason: <c>For</c> rolls to this round, so <c>SetOffLastTurn</c> is
/// exactly the count that stood when the player last passed.
///
/// IT DOES NOT READ KAEYA'S BLIND. That card names Grounded, and a marker that
/// quietly paid a second power would be a rule the player was never shown.
/// </summary>
public sealed class LionsFangPower : PowerModel, ILocalizationProvider
{
    public List<(string, string)>? Localization => new()
    {
        ("title", "Lion's Fang, Fair Protector"),
        ("description",
            "At the start of your turn, if none of your [gold]Bombs[/gold] "
          + "went off last turn, gain [blue]{Amount}[/blue] [gold]Block[/gold] "
          + "and draw 1 card."),
    };

    public override PowerType Type => PowerType.Buff;

    public override PowerStackType StackType => PowerStackType.Counter;

    public override async Task AfterPlayerTurnStart(
        PlayerChoiceContext choiceContext, Player player)
    {
        if (Owner == null || player.Creature != Owner) return;
        if (Amount <= 0) return;
        if (KleeOverhaulLedger.For(Owner).SetOffLastTurn > 0) return;
        await CreatureCmd.GainBlock(Owner, Amount, ValueProp.Unpowered, null);
        // A LITERAL 1, in both engines and for the reason tier0's
        // `MC_LIONS_FANG_DRAW` comment gives: naming it would make
        // `lint_prose_constants` read every "Draw 1 card" in the mod as an
        // un-interpolated copy of this slice's constant. The row's own
        // `description:` is what both engines print.
        await CardPileCmd.Draw(choiceContext, 1, player);
    }
}
